types/ember-data/index.d.ts

```diff
@@ -2,12 +2,26 @@
 // Project: https://github.com/emberjs/data
 // Definitions by: Derek Wickern <https://github.com/dwickern>
 //                 Mike North <https://github.com/mike-north>
+//                 Chris Krycho <https://github.com/chriskrycho>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 2.4
 
-declare module "ember-data" {
-    import Ember from "ember";
+declare module 'ember-data' {
+    import Ember from 'ember';
     import RSVP from 'rsvp';
+
+    export interface ModelRegistry {
+        [key: string]: DS.Model;
```

Comment:
Should we have this by default? If you make a typo in the model name you'll get back a Model instead of getting a compiler error. I think it would be better to have empty registries by default, and the user can add this line to their own code if they want that trade-off (or if they don't want to deal with type registries at all). If we declare it like this, there's no way to opt out of the unsafety.

Reply:
Good point. I was thinking about backwards compatibility to avoid breaking consumers of the types, but if anyone has used the type coercion form at all, this is *already* a breaking change. I'll just add this to the list of things to document in ember-cli-typescript today.

I think what we should do is:

1. For model, adapter, and serializer, do as you suggest and give users a way to opt in.
2. For services and controllers, do *not* supply the string key, but supply these as two different type signatures:

    ```ts
    function controller(): ComputedProperty<Ember.Controller>;
    function controller<K extends keyof ControllerRegistry>(
        name?: K
    ): ComputedProperty<ControllerRegistry[K]>;
    
    function service(): ComputedProperty<Ember.Service>;
    function service<K extends keyof ServiceRegistry>(
        name?: K
    ): ComputedProperty<ServiceRegistry[K]>;
    ```
    
    That way, consumers do get a kind of backwards-compatibility to just get `Service` or `Controller` instances in the current scheme, and they get to opt in simply by supplying the string key.

Reply:
We should discourage the no-argument form, maybe using custom tslint rules?

Reply:
@dwickern that seems like a reasonable idea, though I *suspect* people will mostly figure out to avoid it by dint of "Wait, I can't get this to type check…"